File: Calculator/android-native/app/src/main/java/com/qaantum/calculatorhub/calculators/FinalCalculators.kt
package com.qaantum.calculatorhub.calculators

import kotlin.math.*

// Matrix Determinant Calculator (2x2 and 3x3)
class MatrixDeterminantCalculator {
    fun det2x2(a: Double, b: Double, c: Double, d: Double): Double = a * d - b * c
    
    fun det3x3(matrix: Array<DoubleArray>): Double {
        val a = matrix[0][0]; val b = matrix[0][1]; val c = matrix[0][2]
        val d = matrix[1][0]; val e = matrix[1][1]; val f = matrix[1][2]
        val g = matrix[2][0]; val h = matrix[2][1]; val i = matrix[2][2]
        return a * (e * i - f * h) - b * (d * i - f * g) + c * (d * h - e * g)
    }
}

// Color Converter (RGB/HEX/HSL)
data class ColorResult(val hex: String, val rgb: Triple<Int, Int, Int>, val hsl: Triple<Int, Int, Int>)
class ColorConverter {
    fun rgbToHex(r: Int, g: Int, b: Int): String = "#%02X%02X%02X".format(r, g, b)
    
    fun hexToRgb(hex: String): Triple<Int, Int, Int> {
        val clean = hex.removePrefix("#")
        return try {
            Triple(clean.substring(0, 2).toInt(16), clean.substring(2, 4).toInt(16), clean.substring(4, 6).toInt(16))
        } catch (e: Exception) { Triple(0, 0, 0) }
    }
    
    fun rgbToHsl(r: Int, g: Int, b: Int): Triple<Int, Int, Int> {
        val rf = r / 255.0; val gf = g / 255.0; val bf = b / 255.0
        val max = maxOf(rf, gf, bf); val min = minOf(rf, gf, bf)
        val l = (max + min) / 2
        if (max == min) return Triple(0, 0, (l * 100).toInt())
        val d = max - min
        val s = if (l > 0.5) d / (2 - max - min) else d / (max + min)
        val h = when (max) {
            rf -> ((gf - bf) / d + (if (gf < bf) 6 else 0)) / 6
            gf -> ((bf - rf) / d + 2) / 6
            else -> ((rf - gf) / d + 4) / 6
        }
        return Triple((h * 360).toInt(), (s * 100).toInt(), (l * 100).toInt())
    }
}

// Loan Affordability Calculator
data class LoanAffordabilityResult(val maxLoan: Double, val monthlyPayment: Double, val totalInterest: Double)
class LoanAffordabilityCalculator {
    fun calculate(monthlyIncome: Double, debtToIncomeRatio: Double, rate: Double, termMonths: Int): LoanAffordabilityResult {
        val maxPayment = monthlyIncome * (debtToIncomeRatio / 100)
        val monthlyRate = rate / 100 / 12
        val maxLoan = if (monthlyRate == 0.0) maxPayment * termMonths 
                      else maxPayment * (1 - (1 + monthlyRate).pow(-termMonths)) / monthlyRate
        return LoanAffordabilityResult(maxLoan, maxPayment, maxPayment * termMonths - maxLoan)
    }
}

// Refinance Calculator
data class RefinanceResult(val newPayment: Double, val oldPayment: Double, val monthlySavings: Double, val breakEvenMonths: Int)
class RefinanceCalculator {
    fun calculate(loanBalance: Double, oldRate: Double, newRate: Double, termMonths: Int, closingCosts: Double): RefinanceResult {
        val oldMonthlyRate = oldRate / 100 / 12
        val newMonthlyRate = newRate / 100 / 12
        val oldPayment = loanBalance * (oldMonthlyRate * (1 + oldMonthlyRate).pow(termMonths)) / ((1 + oldMonthlyRate).pow(termMonths) - 1)
        val newPayment = loanBalance * (newMonthlyRate * (1 + newMonthlyRate).pow(termMonths)) / ((1 + newMonthlyRate).pow(termMonths) - 1)
        val savings = oldPayment - newPayment
        val breakEven = if (savings > 0) ceil(closingCosts / savings).toInt() else Int.MAX_VALUE
        return RefinanceResult(newPayment, oldPayment, savings, breakEven)
    }
}

// TVM Calculator (Time Value of Money)
data class TVMResult(val futureValue: Double?, val presentValue: Double?, val payment: Double?, val periods: Int?, val rate: Double?)
class TVMCalculator {
    fun futureValue(pv: Double, rate: Double, periods: Int, pmt: Double = 0.0): Double {
        val r = rate / 100
        return pv * (1 + r).pow(periods) + pmt * ((1 + r).pow(periods) - 1) / r
    }
    fun presentValue(fv: Double, rate: Double, periods: Int, pmt: Double = 0.0): Double {
        val r = rate / 100
        return fv / (1 + r).pow(periods) - pmt * (1 - (1 + r).pow(-periods)) / r
    }
}

// Debt Snowball Calculator
data class DebtInfo(val name: String, val balance: Double, val minPayment: Double, val rate: Double)
data class DebtPayoffResult(val totalMonths: Int, val totalInterest: Double, val payoffOrder: List<String>)
class DebtSnowballCalculator {
    fun calculate(debts: List<DebtInfo>, extraPayment: Double): DebtPayoffResult {
        val sorted = debts.sortedBy { it.balance } // Snowball: smallest first
        var totalMonths = 0
        var totalInterest = 0.0
        val order = mutableListOf<String>()
        var extra = extraPayment
        val balances = sorted.map { it.balance }.toMutableList()
        
        while (balances.any { it > 0 }) {
            totalMonths++
            if (totalMonths > 600) break // Safety limit
            
            for (i in sorted.indices) {
                if (balances[i] <= 0) continue
                val interest = balances[i] * sorted[i].rate / 100 / 12
                totalInterest += interest
                val payment = sorted[i].minPayment + (if (balances.take(i).all { it <= 0 }) extra else 0.0)
                balances[i] = balances[i] + interest - payment
                if (balances[i] <= 0 && !order.contains(sorted[i].name)) {
                    order.add(sorted[i].name)
                }
            }
        }
        return DebtPayoffResult(totalMonths, totalInterest, order)
    }
}

// Amortization Schedule
data class AmortizationRow(val month: Int, val payment: Double, val principal: Double, val interest: Double, val balance: Double)
class AmortizationCalculator {
    fun generate(loanAmount: Double, rate: Double, termMonths: Int): List<AmortizationRow> {
        val monthlyRate = rate / 100 / 12
        val payment = loanAmount * (monthlyRate * (1 + monthlyRate).pow(termMonths)) / ((1 + monthlyRate).pow(termMonths) - 1)
        val schedule = mutableListOf<AmortizationRow>()
        var balance = loanAmount
        
        for (month in 1..termMonths) {
            val interest = balance * monthlyRate
            val principal = payment - interest
            balance -= principal
            schedule.add(AmortizationRow(month, payment, principal, interest, maxOf(0.0, balance)))
        }
        return schedule
    }
}

// Rental Property Calculator
data class RentalResult(val cashFlow: Double, val capRate: Double, val roi: Double, val cashOnCash: Double)
class RentalPropertyCalculator {
    fun calculate(purchasePrice: Double, downPayment: Double, monthlyRent: Double, monthlyExpenses: Double, mortgagePayment: Double): RentalResult {
        val annualIncome = monthlyRent * 12
        val annualExpenses = (monthlyExpenses + mortgagePayment) * 12
        val noi = annualIncome - (monthlyExpenses * 12) // Net Operating Income (before mortgage)
        val cashFlow = monthlyRent - monthlyExpenses - mortgagePayment
        val capRate = (noi / purchasePrice) * 100
        val roi = ((cashFlow * 12) / purchasePrice) * 100
        val cashOnCash = ((cashFlow * 12) / downPayment) * 100
        return RentalResult(cashFlow * 12, capRate, roi, cashOnCash)
    }
}

File: Calculator/ios-native/CalculatorHub/Calculators/FinalCalculators.swift
import Foundation

// Matrix Determinant Calculator
class MatrixDeterminantCalculator {
    func det2x2(_ a: Double, _ b: Double, _ c: Double, _ d: Double) -> Double { a * d - b * c }
    
    func det3x3(_ matrix: [[Double]]) -> Double {
        let a = matrix[0][0]; let b = matrix[0][1]; let c = matrix[0][2]
        let d = matrix[1][0]; let e = matrix[1][1]; let f = matrix[1][2]
        let g = matrix[2][0]; let h = matrix[2][1]; let i = matrix[2][2]
        return a * (e * i - f * h) - b * (d * i - f * g) + c * (d * h - e * g)
    }
}

// Color Converter
struct ColorResult { let hex: String; let rgb: (Int, Int, Int); let hsl: (Int, Int, Int) }
class ColorConverter {
    func rgbToHex(_ r: Int, _ g: Int, _ b: Int) -> String { String(format: "#%02X%02X%02X", r, g, b) }
    
    func hexToRgb(_ hex: String) -> (Int, Int, Int) {
        let clean = hex.hasPrefix("#") ? String(hex.dropFirst()) : hex
        guard clean.count == 6 else { return (0, 0, 0) }
        let r = Int(clean.prefix(2), radix: 16) ?? 0
        let g = Int(clean.dropFirst(2).prefix(2), radix: 16) ?? 0
        let b = Int(clean.suffix(2), radix: 16) ?? 0
        return (r, g, b)
    }
    
    func rgbToHsl(_ r: Int, _ g: Int, _ b: Int) -> (Int, Int, Int) {
        let rf = Double(r) / 255; let gf = Double(g) / 255; let bf = Double(b) / 255
        let maxC = max(rf, gf, bf); let minC = min(rf, gf, bf)
        let l = (maxC + minC) / 2
        guard maxC != minC else { return (0, 0, Int(l * 100)) }
        let d = maxC - minC
        let s = l > 0.5 ? d / (2 - maxC - minC) : d / (maxC + minC)
        var h: Double
        switch maxC {
        case rf: h = ((gf - bf) / d + (gf < bf ? 6 : 0)) / 6
        case gf: h = ((bf - rf) / d + 2) / 6
        default: h = ((rf - gf) / d + 4) / 6
        }
        return (Int(h * 360), Int(s * 100), Int(l * 100))
    }
}

// Loan Affordability Calculator
struct LoanAffordabilityResult { let maxLoan: Double; let monthlyPayment: Double; let totalInterest: Double }
class LoanAffordabilityCalculator {
    func calculate(monthlyIncome: Double, debtToIncomeRatio: Double, rate: Double, termMonths: Int) -> LoanAffordabilityResult {
        let maxPayment = monthlyIncome * (debtToIncomeRatio / 100)
        let monthlyRate = rate / 100 / 12
        let maxLoan = monthlyRate == 0 ? maxPayment * Double(termMonths) : maxPayment * (1 - pow(1 + monthlyRate, Double(-termMonths))) / monthlyRate
        return LoanAffordabilityResult(maxLoan: maxLoan, monthlyPayment: maxPayment, totalInterest: maxPayment * Double(termMonths) - maxLoan)
    }
}

// Refinance Calculator
struct RefinanceResult { let newPayment: Double; let oldPayment: Double; let monthlySavings: Double; let breakEvenMonths: Int }
class RefinanceCalculator {
    func calculate(loanBalance: Double, oldRate: Double, newRate: Double, termMonths: Int, closingCosts: Double) -> RefinanceResult {
        let oldMonthlyRate = oldRate / 100 / 12
        let newMonthlyRate = newRate / 100 / 12
        let oldPayment = loanBalance * (oldMonthlyRate * pow(1 + oldMonthlyRate, Double(termMonths))) / (pow(1 + oldMonthlyRate, Double(termMonths)) - 1)
        let newPayment = loanBalance * (newMonthlyRate * pow(1 + newMonthlyRate, Double(termMonths))) / (pow(1 + newMonthlyRate, Double(termMonths)) - 1)
        let savings = oldPayment - newPayment
        let breakEven = savings > 0 ? Int(ceil(closingCosts / savings)) : Int.max
        return RefinanceResult(newPayment: newPayment, oldPayment: oldPayment, monthlySavings: savings, breakEvenMonths: breakEven)
    }
}

// TVM Calculator
class TVMCalculator {
    func futureValue(pv: Double, rate: Double, periods: Int, pmt: Double = 0) -> Double {
        let r = rate / 100
        return pv * pow(1 + r, Double(periods)) + pmt * (pow(1 + r, Double(periods)) - 1) / r
    }
    func presentValue(fv: Double, rate: Double, periods: Int, pmt: Double = 0) -> Double {
        let r = rate / 100
        return fv / pow(1 + r, Double(periods)) - pmt * (1 - pow(1 + r, Double(-periods))) / r
    }
}

// Amortization Schedule
struct AmortizationRow { let month: Int; let payment: Double; let principal: Double; let interest: Double; let balance: Double }
class AmortizationCalculator {
    func generate(loanAmount: Double, rate: Double, termMonths: Int) -> [AmortizationRow] {
        let monthlyRate = rate / 100 / 12
        let payment = loanAmount * (monthlyRate * pow(1 + monthlyRate, Double(termMonths))) / (pow(1 + monthlyRate, Double(termMonths)) - 1)
        var schedule: [AmortizationRow] = []; var balance = loanAmount
        
        for month in 1...termMonths {
            let interest = balance * monthlyRate
            let principal = payment - interest
            balance -= principal
            schedule.append(AmortizationRow(month: month, payment: payment, principal: principal, interest: interest, balance: max(0, balance)))
        }
        return schedule
    }
}

// Rental Property Calculator
struct RentalResult { let annualCashFlow: Double; let capRate: Double; let roi: Double; let cashOnCash: Double }
class RentalPropertyCalculator {
    func calculate(purchasePrice: Double, downPayment: Double, monthlyRent: Double, monthlyExpenses: Double, mortgagePayment: Double) -> RentalResult {
        let noi = monthlyRent * 12 - monthlyExpenses * 12
        let cashFlow = monthlyRent - monthlyExpenses - mortgagePayment
        let capRate = (noi / purchasePrice) * 100
        let roi = ((cashFlow * 12) / purchasePrice) * 100
        let cashOnCash = ((cashFlow * 12) / downPayment) * 100
        return RentalResult(annualCashFlow: cashFlow * 12, capRate: capRate, roi: roi, cashOnCash: cashOnCash)
    }
}
